### PR TITLE
Behavioural cloning

### PR DESCRIPTION
**Pull request by qxcv**
This PR adds behavioural cloning, and also changes the rollout API to be trajectory-centric. `total_reward` has also been replaced with a function that computes mean return over a series of trajectories, which should provide more a more meaningful indicator of the performance of a policy.

### CODE REVIEW

File: .style.yapf [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[style]
```
[the remark this answers is not in the document]
Comment (qxcv) [reply .style.yapf]: Yes, it's for [this tool](https://github.com/google/yapf). I can remove it if you'd prefer.

File: imitation/util/rollout.py [hunk 2]
```diff
@@ -1,32 +1,81 @@
...
+  """Accumulates trajectories step-by-step. For use in generate() only."""
+
+  def __init__(self):
+    self.partial_trajectories = {}
+
+  def finish_trajectory(self, idx):
+    """Complete the trajectory labelled with `idx`, and return list of
+    completed trajectories popped from self.partial_trajectories."""
+    if idx not in self.partial_trajectories:
+      return []
```
[the remark this answers is not in the document]
Comment (qxcv) [reply imitation/util/rollout.py]: I originally added this to handle the case where the trajectory finishes on the first time step, so there's nothing to add to the buffer (see `trajectories.extend(finish_trajectory())`). That cannot happen with my current code, so I will just remove the branch.

File: imitation/util/rollout.py [hunk 3]
```diff
@@ -62,82 +116,84 @@ def generate(policy, env, *, n_timesteps=None, n_episodes=None,
...
-    n_steps = len(rollout_obs_new)
+    zip_iter = enumerate(
+        zip(obs_old_batch, act_batch, obs_batch, rew_batch, done_batch))
+    for idx, (obs_old, act, obs, rew, done) in zip_iter:
+      if done:
+        # finish idx-th trajectory
+        # FIXME: this will break horribly if a trajectory ends after the first
+        # action, b/c the trajectory will consist of just a single obs. The
+        # "correct" fix for this is to PATCH STABLE BASELINES SO THAT ITS
+        # VECENV GIVES US A CORRECT FINAL OBSERVATION TO ADD.
```
[the remark this answers is not in the document]
Comment (qxcv) [reply imitation/util/rollout.py]: Yes, that's a good idea. IMO this is really important, so for now I'll add an exception admonishing us to solve #1 (maybe add to Adam's fork & then make a PR?).ad

**Comment by qxcv** [comment]
Changes added. Thanks for making the docstring changes!

**Comment by shwang** [comment]
image

**Pull request merged**